src/runtime2/loader/dom-ready.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  CanRender,
+  contextProp,
+  removeProp,
+  setProp,
+  subscribe,
+  unsubscribe,
+  useSyncEffect,
+} from '../../context';
+import {hasNextNodeInDocumentOrder} from '../../dom';
+
+/** @type {!ContextProp<?function(!Element)>} */
+// eslint-disable-next-line local/no-export-side-effect
+export const DomReadyProp = contextProp('DomReady', {
+  needsParent: true,
+  defaultValue: null,
+});
+
+/**
+ * @param {!Node} root
+ * @param {boolean} treeReady
+ */
+export function DomReadyProvider(root, treeReady) {
+  useSyncEffect(() => {
+    if (treeReady) {
+      return;
+    }
+
+    let pending = [];
+
+    const checkPending = () => {
+      for (let i = 0; i < pending.length; i++) {
+        const node = pending[i];
+        if (hasNextNodeInDocumentOrder(node, root)) {
+          // Remove resource before build to remove it from the pending list
+          // in either case the build succeed or throws an error.
+          pending.splice(i--, 1);
+          removeProp(node, CanRender, DomReadyProvider);
+        }
+      }
+    };
+
+    /**
+     * @param {!Node} node
+     */
+    const onReady = (node) => {
+      if (!pending) {
```

Comment:
> How could the pending=null case ever happen at the time of an onReady callback? Is it a race condition with the cleanup?

Yes. This is a somewhat rare case of a race condition in this API. `onReady` prop is recomputed and notified over some cycles, so the old value might still be used. I'm not super happy about it. However, it's not too bad because the nuance of the race condition is fully encapsulated and fully handled in this one component. Still I'm looking if I can find something more natural.

> onReady Doesn't feel like the right name, since most of the "ready" referred to in this function is when the dom is ready. this ready refers to when the component in question appears (but is blocked on dom readiness)

Hmm. Here's how I read it: the DOM node passed to `onReady` is fully parsed. And the whole feature is basically "Block rendering of this element until it's fully parsed". These names are indeed somewhat confusing. I'll see if something else would work better.